tiledb/api/src/query/read/aggregate.rs

```diff
@@ -0,0 +1,665 @@
+            }
+            AggregateEnumBuilder::UInt64(agg_builder) => {
+                AggregateEnumReader::UInt64(agg_builder.build())
+            }
+            AggregateEnumBuilder::Int8(agg_builder) => {
+                AggregateEnumReader::Int8(agg_builder.build())
+            }
+            AggregateEnumBuilder::Int16(agg_builder) => {
+                AggregateEnumReader::Int16(agg_builder.build())
+            }
+            AggregateEnumBuilder::Int32(agg_builder) => {
+                AggregateEnumReader::Int32(agg_builder.build())
+            }
+            AggregateEnumBuilder::Int64(agg_builder) => {
+                AggregateEnumReader::Int64(agg_builder.build())
+            }
+            AggregateEnumBuilder::Float32(agg_builder) => {
+                AggregateEnumReader::Float32(agg_builder.build())
+            }
+            AggregateEnumBuilder::Float64(agg_builder) => {
+                AggregateEnumReader::Float64(agg_builder.build())
+            }
+        }
+    }
+}
+
+impl<Q, T> Query for AggregateTypedReader<Q, T>
+where
+    Q: Query,
+{
+    fn base(&self) -> &QueryBase {
+        self.base.base()
+    }
+
+    fn finalize(self) -> TileDBResult<Array>
+    where
+        Self: Sized,
+    {
+        self.base.finalize()
+    }
+}
+
+impl<Q> Query for AggregateEnumReader<Q>
+where
+    Q: Query,
+{
+    fn base(&self) -> &QueryBase {
+        agg_enum_reader_go!(self, _DT, reader, reader.base.base())
+    }
+
+    fn finalize(self) -> TileDBResult<Array>
+    where
+        Self: Sized,
+    {
+        agg_enum_reader_go!(self, _DT, reader, reader.base.finalize())
+    }
+}
+
+impl<Q, T> ReadQuery for AggregateTypedReader<Q, T>
+where
+    Q: ReadQuery,
+    T: Copy,
+{
+    type Intermediate = ();
+    type Final = (T, Q::Final);
+
+    fn step(
+        &mut self,
+    ) -> TileDBResult<ReadStepOutput<Self::Intermediate, Self::Final>> {
+        // Register the data buffer (set data buffer)
+        let context = self.base().context();
+        let cquery = **self.base().cquery();
+        let location_ptr = &mut self.data as *mut T;
+        let c_bufptr = location_ptr as *mut std::ffi::c_void;
+        let c_sizeptr = &mut self.data_size as *mut u64;
+        let agg_str: &CString = &self.agg_str;
+        let agg_c_ptr = agg_str.as_c_str().as_ptr();
+
+        context.capi_call(|ctx| unsafe {
+            ffi::tiledb_query_set_data_buffer(
+                ctx, cquery, agg_c_ptr, c_bufptr, c_sizeptr,
+            )
+        })?;
+
+        let base_result = self.base.step()?;
+
+        // There are no intermediate results for aggregates since the buffer size should be one
+        // element (and therefore, no space constraints).
+        let (return_val, base_q) = match base_result {
+            ReadStepOutput::Final(base_q) => (self.data, base_q),
+            ReadStepOutput::Intermediate(_) => {
+                unreachable!("Aggregate step function.")
```

Comment:
This doesn't work since the base result type is not required to implement debug or display. How do we mitigate this problem?

Reply:
`ReadStepOutput` derives Debug, so it should be Debug when its type parameters are.

As far as this `unerachable!` goes, that's a fair point. But I think it's ok to put the needed bounds in there.
- the bound on `T` should be PhysicalType rather than Copy, and thus implements Debug. IIRC I left this comment previously about a different bound but it applies here too.
- `<Q as ReadQuery>::Intermediate` and `<Q as ReadQuery>::Final` are always going to be Debug in practice anyway since aggregates can only be put on top of each other and the base query.